geomesa-arrow/geomesa-arrow-gt/src/main/scala/org/locationtech/geomesa/arrow/io/DeltaWriter.scala

```diff
@@ -564,44 +555,54 @@ object DeltaWriter extends StrictLogging {
       dictionaryFields: Seq[String],
       deltas: Array[Array[Array[Byte]]],
       encoding: SimpleFeatureEncoding): MergedDictionaries = {
-    import org.locationtech.geomesa.utils.conversions.ScalaImplicits.{RichArray, RichTraversableOnce}
+    import org.locationtech.geomesa.utils.conversions.ScalaImplicits.RichArray
 
     val allocator = ArrowAllocator("merge-dictionaries")
 
     if (dictionaryFields.isEmpty) {
       return MergedDictionaries(Map.empty, Map.empty, allocator)
     }
 
+    // calculate our vector bindings/metadata once up front
+    val vectorMetadata = dictionaryFields.toArray.map { name =>
+      val descriptor = sft.getDescriptor(name)
+      val bindings = ObjectType.selectType(descriptor)
+      val metadata = Map(SimpleFeatureVector.DescriptorKey -> SimpleFeatureTypes.encodeDescriptor(sft, descriptor))
+      val factory = VectorFactory(allocator)
+      (name, bindings, metadata, factory)
+    }
+
     // create a vector for each dictionary field
     def createNewVectors: Array[ArrowAttributeReader] = {
-      val builder = Array.newBuilder[ArrowAttributeReader]
-      builder.sizeHint(dictionaryFields.length)
-      dictionaryFields.foreach { f =>
-        val descriptor = sft.getDescriptor(f)
+      vectorMetadata.map { case (name, bindings, metadata, factory) =>
         // use the writer to create the appropriate child vector
-        val vector = ArrowAttributeWriter(sft, descriptor, None, encoding, allocator).vector
-        builder += ArrowAttributeReader(descriptor, vector, None, encoding)
+        val vector = ArrowAttributeWriter(name, bindings, None, metadata, encoding, factory).vector
+        ArrowAttributeReader(bindings, vector, None, encoding)
       }
-      builder.result
     }
 
     // final results
     val results = createNewVectors
 
     // re-used queue, gets emptied after each dictionary field
     // [(dictionary value, batch index, index of value in the batch)]
-    val queue = scala.collection.mutable.PriorityQueue.empty[(AnyRef, Int, Int)](queueOrdering)
+    // TODO ordering reverse?
+    val queue = new PriorityQueue[DictionaryMerger](Ordering.ordered[DictionaryMerger])
 
     // merge each threaded delta vector into a single dictionary for that thread
     // Array[(dictionary vector, transfers to result, batch delta mappings)]
-    val allMerges: Array[(Array[ArrowAttributeReader], Array[TransferPair], Array[HashBiMap[Integer, Integer]])] = deltas.map { deltas =>
+//    val allMerges: Array[(Array[ArrowAttributeReader], Array[TransferPair], Array[HashBiMap[Integer, Integer]])] = deltas.map { deltas =>
+    val allMerges: Array[DictionaryMerger] = deltas.mapWithIndex { case (deltas, batch) =>
```

Comment:
I know I don't like zipWithIndex since it allocates a bunch of extra tuples, etc.

I imagine there aren't too many batches here, so it isn't an 'inner loop' situation.  That said, would it be better to have a counter and increment?

Also, this anonymous function is huge.  Is there anyway to break it up sensibly?  Or should it be a do-all function in order to optimize things?

Reply:
`mapWithIndex` is syntactic sugar we wrote that increments a var: https://github.com/locationtech/geomesa/blob/master/geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/conversions/ScalaImplicits.scala#L33-L40
but yeah, this method is large...